cmd/generic-handlers.go

```diff
@@ -65,6 +65,48 @@ func (h requestSizeLimitHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
 	h.handler.ServeHTTP(w, r)
 }
 
+const (
+	maxHeaderSize   = 8 * 1024
+	maxUserDataSize = 2 * 1024
+)
+
+type requestHeaderSizeLimitHandler struct {
+	handler              http.Handler
+	maxSize, maxUserSize int
+}
+
+func setRequestHeaderSizeLimitHandler(h http.Handler) http.Handler {
+	return requestHeaderSizeLimitHandler{handler: h, maxSize: maxHeaderSize, maxUserSize: maxUserDataSize}
+}
+func (h requestHeaderSizeLimitHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if err := h.checkHTTPHeaderSize(r.Header); err != nil {
+		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
+		return
+	}
+	h.handler.ServeHTTP(w, r)
+}
+
+func (h requestHeaderSizeLimitHandler) checkHTTPHeaderSize(header http.Header) error {
```

Comment:
checkHTTPHeaderSize need not be a method on requestHeaderSizeLimitHandler, it can be an independent func.
If you rename it to isHTTPHeaderSizeTooLarge, then it can just return `bool` and you dont need to introduce the error errMetadataTooLarge